[RDY] api: Do not truncate errors <1 MB

justinmk opened the pull request:
Replaces #6559, which was accidentally pushed to neovim repo rather than my fork.

- [x] pass `ErrorType` explicitly to `api_set_error`
- [x] do not translate/localize API errors
- [x] remove `Error.set` field, check `Error.type == kErrorTypeNone` instead
- [x] implement `api_clear_error(Error *)`

sander2 committed:
api: Do not truncate errors <1 MB. #6237

Closes #5984

justinmk committed:
test: api: Do not truncate errors <1 MB.

justinmk commented:
The quickbuild failure is weird:

```
14:15:27,800 INFO  - ok 2008 - terminal window highlighting bold, italics and underline will pass the corresponding attributes on scrollback
14:15:27,800 INFO  - not ok 2009 - terminal window highlighting with custom palette before_each
14:15:27,800 INFO  - # test/functional/terminal/highlight_spec.lua @ 119
14:15:27,800 INFO  - # Failure message: ./test/functional/helpers.lua:140: ./test/functional/ui/screen.lua:320: Add Screen:_handle_cursor_goto
14:15:27,800 INFO  - # stack traceback:
14:15:27,800 INFO  - #     ./test/functional/helpers.lua:140: in function 'run'
14:15:27,800 INFO  - #     ./test/functional/ui/screen.lua:280: in function 'wait'
14:15:27,800 INFO  - #     ./test/functional/ui/screen.lua:215: in function 'expect'
14:15:27,800 INFO  - #     test/functional/terminal/highlight_spec.lua:136: in function <test/functional/terminal/highlight_spec.lua:119>
14:15:27,800 INFO  - #
```

oni-link commented on src/nvim/terminal.c:
```diff
@@ -1217,12 +1218,14 @@ static bool is_focused(Terminal *term)
+    api_free_error(err); \
     if (o.type == kObjectTypeNil) { \
       o = dict_get_value(&globvardict, cstr_as_string(k), &err); \
+      api_free_error(err); \
```
This call can result in a double free:
If the key is not found in `curbuf->b_vars` then `err` is set and freed by the previous call to `api_free_error()`.
But `err->msg` still points to the freed memory.
If the key is found in `globalvardict` then no error is set but an invalid memory address is freed.

justinmk replied:
@oni-link `api_clear_error(Error *)` sets msg to NULL now, so the double-free should be avoided. Thanks for finding that.

oni-link commented on src/nvim/api/private/helpers.c:
```diff
@@ -798,6 +797,11 @@ void api_free_dictionary(Dictionary value)
   xfree(value.items);
 }
 
+void api_free_error(Error value)
```
Should we also use a macro version (like api_set_error()) that checks if an error was set before calling this function? In most cases no error should be set.

justinmk replied:
I added a check in api_clear_error:

```
if (!value->set) {
  return;
}
```

oni-link commented on src/nvim/api/private/helpers.c:
```diff
@@ -965,3 +969,22 @@ static void set_option_value_err(char *key,
   }
 }
+
+void _api_set_error(Error *err, ErrorType errType, const char *format, ...)
```
Could the macro name use the leading underscore instead of this function?

ZyX-I commented on src/nvim/api/private/helpers.h:
```diff
@@ -8,15 +8,10 @@
-    (err)->set = true; \
-    (err)->type = kErrorType##errtype; \
-  } while (0)
+  _api_set_error(err, kErrorType##errtype, __VA_ARGS__)
```
The macro should be all-caps. So the function could be used without leading underscore. And there is no need to duplicate declaration, macro is not used before including `api/private/helpers.h.generated.h`.

ZyX-I replied:
Though I do not see much sense in the macro any longer: it only hides `kErrorType` prefix, not something I would consider good under this circumstances (when passing e.g. Validation to a function you would not expect that it is not a full identifier, but a part of it: all-lower does not look like a macro; with all-caps this is slightly better but I do not think ten characters are worth it).

oni-link commented on src/nvim/api/private/defs.h:
```diff
@@ -38,7 +38,7 @@ typedef enum {
   ErrorType type;
-  char msg[1024];
+  char *msg;
   bool set;
```
Do we still need this flag or is checking `msg != NULL` enough?

ZyX-I replied:
I usually put such data into `type` if present: if I was writing this struct there would be `kErrorTypeNone` and no flag.

justinmk replied:
Added kErrorTypeNone and removed `set`.

oni-link commented on src/nvim/eval.c:
```diff
@@ -16531,7 +16533,7 @@ static void f_termopen(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+  Error err = ERROR_INIT;
   dict_set_var(curbuf->b_vars, cstr_as_string("terminal_job_id"),
                INTEGER_OBJ(rettv->vval.v_number), false, false, &err);
   dict_set_var(curbuf->b_vars, cstr_as_string("terminal_job_pid"),
```
Here `err` could be set two times. After the second "set" we would have a memory leak.

oni-link commented on src/nvim/msgpack_rpc/channel.c:
```diff
@@ -441,6 +441,7 @@ static void handle_request(Channel *channel, msgpack_object *request)
   } else {
     multiqueue_put(channel->events, on_request_event, 1, event_data);
   }
+  api_free_error(error);
```
At this point no error was set, so this call could be removed.

oni-link commented on src/nvim/tui/input.c:
```diff
@@ -225,7 +225,7 @@ static int get_key_code_timeout(void)
   if (nvim_get_option(cstr_as_string("ttimeout"), &err).data.boolean) {
     ms = nvim_get_option(cstr_as_string("ttimeoutlen"), &err).data.integer;
```
In case of an error `....data.integer` seems to be `0`. But should `ms` not keep the value `-1`?

justinmk replied:
@oni-link see https://github.com/neovim/neovim/pull/6569/commits/cfe2a976557e233ec01893ed687751225501d872

justinmk committed:
api_clear_error()

justinmk committed:
api_clear_error: Skip if error was not set.

justinmk committed:
api/internal: Remove `set` field from Error type.

justinmk committed:
api_set_error(): rename

justinmk committed:
tui/input.c: Use default 'ttimeoutlen' if option get fails.

Should never happen, but better to be explicit.

Helped-by: oni-link <knil.ino@gmail.com>

justinmk committed:
api: Do not translate error messages.

Also re-word some error messages:
- "Key does not exist: %s"
- "Invalid channel: %<PRIu64>"
- "Request array size must be 4 (request) or 3 (notification)"
- "String cannot contain newlines"

References #6150

justinmk commented:
Should be ready after CI.

ZyX-I commented on src/nvim/api/buffer.c:
```diff
@@ -317,7 +317,8 @@ void nvim_buf_set_lines(uint64_t channel_id,
       if (l.data[j] == '\n' && channel_id != INTERNAL_CALL) {
-        api_set_error(err, Exception, _("string cannot contain newlines"));
+        api_set_error(err, kErrorTypeException,
+                      "string cannot contain newlines");
```
All above messages start with a capital letter.

ZyX-I commented on src/nvim/api/private/helpers.c:
```diff
@@ -258,30 +259,30 @@ Object get_option_from(void *from, int type, String name, Error *err)
       api_set_error(err,
-                    Exception,
+                    kErrorTypeException,
                     _("Cannot unset option \"%s\" "
```
This is still translated.

ZyX-I commented on src/nvim/api/private/helpers.c:
```diff
@@ -801,6 +802,16 @@ void api_free_dictionary(Dictionary value)
   xfree(value.items);
 }
 
+void api_clear_error(Error *value)
```
FUNC_ATTR_NONNULL_ALL

ZyX-I commented on src/nvim/api/private/helpers.c:
```diff
@@ -965,6 +976,25 @@ static void set_option_value_err(char *key,
   }
 }
+
+void api_set_error(Error *err, ErrorType errType, const char *format, ...)
```
FUNC_ATTR_NONNULL_ALL

ZyX-I commented on src/nvim/api/vim.c:
```diff
@@ -745,38 +745,38 @@ Array nvim_call_atomic(uint64_t channel_id, Array calls, Error *err)
-                    Validation,
-                    _("name must be String"));
+                    kErrorTypeValidation,
+                    "name must be String");
```
Not capital.

ZyX-I commented on src/nvim/api/vim.c:
```diff
@@ -745,38 +745,38 @@ Array nvim_call_atomic(uint64_t channel_id, Array calls, Error *err)
-                    Validation,
-                    _("args must be Array"));
+                    kErrorTypeValidation,
+                    "args must be Array");
```
Not capital.

justinmk commented:
updated
